File: leetcode/Binary_Search_Tree/0173_Binary_Search_Tree_Iterator.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class BSTIterator {
private:
	stack<TreeNode*> nextNodes;
	void pushAll(TreeNode* node) {
		while (node != NULL) {
			nextNodes.emplace(node);
			node = node -> left;
		}
	}
public:
    BSTIterator(TreeNode *root) {
    	pushAll(root);
    }

    /** @return whether we have a next smallest number */
    bool hasNext() {
        return !nextNodes.empty();
    }

    /** @return the next smallest number */
    int next() {
        TreeNode *current = nextNodes.top();
        nextNodes.pop();
        pushAll(current -> right);
        return current -> val;
    }
};

/**
 * Your BSTIterator will be called like this:
 * BSTIterator i = BSTIterator(root);
 * while (i.hasNext()) cout << i.next();
 */

File: leetcode/Binary_Search_Tree/0173_Binary_Search_Tree_Iterator.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator(object):

    def __init__(self, root):
        """
        :type root: TreeNode
        """
        self.stack = []
        self._push_all(root)

    def _push_all(self, node):
        while node:
            self.stack.append(node)
            node = node.left

    # return next smallest num
    def next(self):
        """
        :rtype: int
        """
        current = self.stack.pop()
        self._push_all(current.right)
        return current.val

    def hasNext(self):
        """
        :rtype: bool
        """
        return bool(self.stack)


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()
